---
title: "Analyzing SCLC scATAC-seq from Debbie (Sage lab)"
output: html_document
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'
---
  
```{r init, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = '/Users/smgroves/Documents/Github/multiome-analysis/')
```

For this tutorial, we will be analyzing a single-cell ATAC-seq data set from Debbie in Julien Sage's lab. She generated two RDS files: sclc (all tumor cells) and sclc_NE (only NE subpopulations). This notebook is copied from example.Rmd (PBMC vignette from Signac Github).

First load in Signac, Seurat, and some other packages we will be using for
analyzing mouse data.

```{r setup, message=FALSE}
library(Signac)
library(Seurat)
library(GenomeInfoDb)
library(EnsDb.Mmusculus.v79) 
library(ggplot2)
library(patchwork)
library(dplyr)
set.seed(1234)

```

## Pre-processing workflow from Vignette

When pre-processing chromatin data, Signac uses information from two related
input files, both of which can be created using CellRanger:
  
  * **Peak/Cell matrix**. This is analogous to the gene expression count matrix used to analyze single-cell RNA-seq. However, instead of genes, each row of the matrix represents a region of the genome (a peak), that is predicted to represent a region of open chromatin. Each value in the matrix represents the number of Tn5 integration sites for each single barcode (i.e. a cell) that map within each peak. You can find more detail on the [10X Website](https://support.10xgenomics.com/single-cell-atac/software/pipelines/latest/output/matrices).

  * **Fragment file**. This represents a full list of all unique fragments across all single cells. It is a substantially larger file, is slower to work with, and is stored on-disk (instead of in memory). However, the advantage of retaining this file is that it contains all fragments associated with each single cell, as opposed to only fragments that map to peaks. More information about the fragment file can be found on the 10x Genomics [website](https://support.10xgenomics.com/single-cell-atac/software/pipelines/latest/output/fragments) or on the [sinto website](https://timoast.github.io/sinto/basic_usage.html#create-scatac-seq-fragments-file).

We start by creating a Seurat object using the peak/cell matrix and cell
metadata generated by `cellranger-atac`, and store the path to the fragment
file on disk in the Seurat object:
  
```{r message=FALSE, warning=FALSE, include = FALSE}
# counts <- Read10X_h5(filename = "../vignette_data/atac_v1_sclc_10k_filtered_peak_bc_matrix.h5")
# metadata <- read.csv(
#   file = "../vignette_data/atac_v1_sclc_10k_singlecell.csv",
#   header = TRUE,
#   row.names = 1
# )
# 
# chrom_assay <- CreateChromatinAssay(
#   counts = counts,
#   sep = c(":", "-"),
#   genome = 'hg19',
#   fragments = '../vignette_data/atac_v1_sclc_10k_fragments.tsv.gz',
#   min.cells = 10,
#   min.features = 200
# )
# 
# sclc <- CreateSeuratObject(
#   counts = chrom_assay,
#   assay = "peaks",
#   meta.data = metadata
# )
```

```{r}
sclc <- readRDS('/Users/smgroves/Box/multiome_data/M1/outs/TKO.rds')
sclc
```

## Assays present
1. **RNA**: This is the unprocessed RNA data. It was preprocessed and transformed using SCT as shown here: https://satijalab.org/signac/articles/pbmc_multiomic.html to generate the **"SCT"** assay. 

For our data, Assay data with 32285 features for 8483 cells
First 10 features:
 Xkr4, Gm1992, Gm19938, Gm37381, Rp1, Sox17, Gm37587, Gm37323, Mrpl15, Lypla1 
 
2. **ATAC**: this is the unprocessed ATAC data a (ChromatinAssay) with counts, similar to RNA-seq counts. It comes from the fragments file in Box, and includes Motif information and annotations that Debbie added. 

3. **Peaks**: The set of peaks identified using Cellranger often merges distinct peaks that are close together. This can create a problem for certain analyses, particularly motif enrichment analysis and peak-to-gene linkage. To identify a more accurate set of peaks, we can call peaks using MACS2 with the CallPeaks() function. Here we call peaks on all cells together, but we could identify peaks for each group of cells separately by setting the group.by parameter, and this can help identify peaks specific to rare cell populations. (from https://satijalab.org/signac/articles/pbmc_muliomic.html under **Peak calling**). 
  - For our data: ChromatinAssay data with 200034 features for 8483 cells
  - Variable features: 199464 
  - Genome: 
  - Annotation present: TRUE 
  - Motifs present: FALSE 
  - Fragment files: 1 
  
4. **chromvar**: Chromatin variance: per cell motif activity score from chromVAR (https://greenleaflab.github.io/chromVAR/index.html), shown in the vignette: https://satijalab.org/signac/articles/motif_vignette.html. This allows us to visualize motif activities per cell, and also provides an alternative method of identifying differentially-active motifs between cell types. 

In our case, this is assay data with 633 features for 8483 cells
First 10 features:
 MA0030.1, MA0031.1, MA0051.1, MA0057.1, MA0059.1, MA0066.1, MA0069.1,
MA0070.1, MA0071.1, MA0072.1 

## Normalization and linear dimensional reduction

* Normalization: Signac performs term frequency-inverse document frequency (TF-IDF) normalization. This is a two-step normalization procedure, that both normalizes across cells to correct for differences in cellular sequencing depth, and across peaks to give higher values to more rare peaks.

* Feature selection: The low dynamic range of scATAC-seq data makes it challenging to perform variable feature selection, as we do for scRNA-seq. Instead, we can choose to use only the top _n_% of features (peaks) for dimensional reduction, or remove features present in less than _n_ cells with the `FindTopFeatures()` function. Here, we will all features, though we note that we see very similar results when using only a subset of features (try setting min.cutoff to 'q75' to use the top 25% all peaks), with faster runtimes. Features used for dimensional reduction are automatically set as `VariableFeatures()` for the Seurat object by this function.

* Dimension reduction: We next run singular value decomposition (SVD) on the TD-IDF matrix, using the features (peaks) selected above. This returns a reduced dimension representation of the object (for users who are more familiar with scRNA-seq, you can think of this as analogous to the output of PCA).

The combined steps of TF-IDF followed by SVD are known as **latent semantic
indexing (LSI)**, and were first introduced for the analysis of scATAC-seq data by
[Cusanovich et al. 2015](https://science.sciencemag.org/content/367/6473/45.full).

```{r message=FALSE, warning=FALSE}
#Debbie did this in her Rmd file
DefaultAssay(sclc) <- "peaks"

# sclc <- RunTFIDF(sclc)
# sclc<- FindTopFeatures(sclc, min.cutoff = 5)
# sclc <- RunSVD(sclc)
```

The first LSI component often captures sequencing depth (technical variation)
rather than biological variation. If this is the case, the component should be
removed from downstream analysis. We can assess the correlation between each LSI
component and sequencing depth using the `DepthCor()` function:

```{r}
DepthCor(sclc)
```

Here we see there is a very strong correlation between the first LSI component
and the total number of counts for the cell, so we will perform downstream steps
without this component.

## Non-linear dimension reduction and clustering

Now that the cells are embedded in a low-dimensional space, we can use methods
commonly applied for the analysis of scRNA-seq data to perform graph-based
clustering and non-linear dimension reduction for visualization. The functions
`RunUMAP()`, `FindNeighbors()`, and `FindClusters()` all come from the Seurat
package.

### UMAP on ATAC data

The UMAP on just ATAC data is very similar to the full UMAP, below.
```{r message=FALSE, warning=FALSE}
#dimentionality reduction of ATAC data <- From Debbie's Rmd file***
sclc <- RunUMAP(object = sclc, reduction = 'lsi', dims = 2:30)
sclc <- FindNeighbors(object = sclc, reduction = 'lsi', dims = 2:30)
sclc <- FindClusters(object = sclc, verbose = FALSE, algorithm = 2, resolution = 0.5)
DimPlot(object = sclc, label = TRUE, reduction = 'umap') + NoLegend()
# ^^ Umap on ATAC: name = UMAP
```
## Joint UMAP

```{r message=FALSE, warning=FALSE}
#RNA
DefaultAssay(sclc) <- "RNA"
# store mitochondrial percentage in object meta data
sclc <- PercentageFeatureSet(sclc, pattern = "mt-", col.name = "percent.mt")

# run sctransform
# BiocManager::install("glmGamPoi")
library(glmGamPoi)
sclc <- SCTransform(sclc, vars.to.regress = "percent.mt", verbose = FALSE,method = "glmGamPoi")

# These are now standard steps in the Seurat workflow for visualization and clustering
sclc <- RunPCA(sclc, verbose = FALSE)
sclc <- RunUMAP(sclc, dims = 1:30, verbose = FALSE)

sclc <- FindNeighbors(sclc, dims = 1:30)
sclc <- FindClusters(sclc)

# build a joint neighbor graph using both assays
sclc <- FindMultiModalNeighbors(
  object = sclc,
  reduction.list = list("pca", "lsi"),
  dims.list = list(1:30, 2:30),
  modality.weight.name = "RNA.weight",
  verbose = TRUE
)

# build a joint UMAP visualization
sclc <- RunUMAP(
  object = sclc,
  nn.name = "weighted.nn",
  assay = "SCT",
  verbose = TRUE,
  reduction.key = 'UMAP_joint',
  reduction.name = 'UMAP_joint'
)

DimPlot(sclc, label = TRUE, repel = TRUE, reduction = "UMAP_joint") + NoLegend()

```

## Finding cluster markers using chromVAR

To find differentially accessible regions between clusters of cells, we can
perform a differential accessibility (DA) test. We utilize logistic regression
for DA, as suggested by
[Ntranos et al. 2018](https://www.biorxiv.org/content/10.1101/258566v2)
for scRNA-seq data, and add the total number of fragments as a latent variable
to mitigate the effect of differential sequencing depth on the result.
For sparse data (such as scATAC-seq), we find it is often necessary
to lower the `min.pct` threshold in `FindMarkers()` from the default (0.1, which
was designed for scRNA-seq data). Here we
will focus on comparing Naive CD4 cells and CD14 monocytes, but any groups of
cells can be compared using these methods. We can also visualize these marker
peaks on a violin plot, feature plot, dot plot, heat map, or any
[visualization tool in Seurat](https://satijalab.org/seurat/v3.0/visualization_vignette.html).

```{r chromvar-markers}
# From Debbie
#finding specific markers
#  DefaultAssay(sclc) <- "chromvar"
# 
# # find markers for every cluster compared to all remaining cells, report only the positive
# # ones
# sclc_motif_markers <- FindAllMarkers(sclc, only.pos = TRUE, min.pct = 0.05, logfc.threshold = 0.5)
# sclc_motif_markers %>%
#     group_by(cluster) %>%
#     slice_max(n = 2, order_by = avg_log2FC)
# 
# #Doing this just to get a full list of motif names
# da_peaks <- FindMarkers(
#   object = sclc,
#   ident.1 = 1,
#   only.pos = TRUE,
#   test.use = 'LR',
#   min.pct = 0.05,
#   latent.vars = 'nCount_peaks'
# )
# 
#  top.da.peak <- rownames(da_peaks[da_peaks$p_val < 0.005, ])
# # test enrichment
#  
# enriched.motifs <- FindMotifs(
#   object = sclc,
#   features = top.da.peak
# )
sclc_motifs_all <- read.csv('/Users/smgroves/Box/multiome_data/M1/outs/sclc_motifs.all.csv', row.names = 1)

# shared_motifs = enriched.motifs[enriched.motifs$motif %in% sclc_motifs_all$gene,]
# colnames(sclc_motifs_all)[8]= "motif"
# sclc_motifs_labelled = merge( sclc_motifs_all, shared_motifs, by = "motif")
# sclc_motifs_labelled =sclc_motifs_labelled[,-c(9:13)]
# sclc_motifs_labelled =sclc_motifs_labelled[,-c(2,9)]
sclc_motifs_labelled <- read.csv('/Users/smgroves/Box/multiome_data/M1/outs/sclc_motifs_labelled.csv', row.names = 1)
sclc_motifs_labelled[1:5,]
```

## Finding cluster markers using RNA
```{r}
# From Debbie's Rmd file

DefaultAssay(sclc) <- "SCT"
features_list <- c("Lyz2", "Calca", "Sftpc","Scgb1a1","Cyp2s1","Ncam1","Ddc","Ascl1","Chga")
DotPlot(sclc, features = "Gata1") + RotatedAxis()
DotPlot(sclc, features = features_list) + RotatedAxis()

```

```{r rna-markers}
#Find markers RNA
# find markers for every cluster compared to all remaining cells, report only the positive
# ones
# sclc.markers_genes <- FindAllMarkers(sclc, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
# sclc.markers_genes %>%
#     group_by(cluster) %>%
#     slice_max(n = 2, order_by = avg_log2FC)
sclc.marker.genes <- read.csv('/Users/smgroves/Box/multiome_data/M1/outs/sclc_marker_genes.csv', row.names = 1)
FeaturePlot(
  object = sclc,
  features = c('Ascl1', 'Nfib', 'Yap1', 'Rest', 'Hes1'),
  pt.size = 0.1,
  max.cutoff = 'q95',
  ncol = 3,
  reduction = 'UMAP_joint'
)
```

```{r cell-cycle}
#Cell cycle score
# A list of cell cycle markers, from Tirosh et al, 2015, is loaded with Seurat.  We can
# segregate this list into markers of G2/M phase and markers of S phase

# s.genes <- cc.genes$s.genes
# g2m.genes <- cc.genes$g2m.genes
# 
# sclc <- CellCycleScoring(sclc, s.features = s.genes_1, g2m.features = g2m.genes_1, set.ident = TRUE)
# 
# # view cell cycle scores and phase assignments
# head(sclc[[]])
# sclc <- ScaleData(sclc, features = g2m.genes_1)
# sclc <- ScaleData(sclc, features = s.genes_1)
# 
# sclc <- RunPCA(sclc, features = c(s.genes_1, g2m.genes_1))
# DimPlot(sclc,reduction = "umap")
# 
# sclc$CC.Difference <- sclc$S.Score - sclc$G2M.Score
# sclc <- ScaleData(sclc, vars.to.regress = "CC.Difference", features = rownames(sclc))
# 
# Idents(sclc) = "seurat_clusters"

Cell_phase = table(Idents(sclc),sclc$Phase) 
Cell_phase

```


## Create a gene activity matrix --> Debbie plans to do this and update the RDS file.

The UMAP visualization reveals the presence of multiple cell groups in human 
blood. If you are familiar with our
[scRNA-seq analyses of sclc](https://satijalab.org/seurat/v3.0/sclc3k_tutorial.html),
you may even recognize the presence of certain myeloid and lymphoid populations
in the scATAC-seq data. However, annotating and interpreting clusters is more
challenging in scATAC-seq data as much less is known about the functional roles
of noncoding genomic regions than is known about protein coding regions (genes). 

However, we can try to quantify the activity of each gene in the genome by
assessing the chromatin accessibility associated with each gene, and create a
new gene activity assay derived from the scATAC-seq data. Here we will use a
simple approach of summing the fragments intersecting the gene body and promoter
region (we also recommend exploring the
[Cicero](https://cole-trapnell-lab.github.io/cicero-release/) tool, which can
accomplish a similar goal, and we provide a vignette showing how to run Cicero
within a Signac workflow [here](cicero.html)). 

To create a gene activity matrix, we extract gene coordinates and extend them
to include the 2 kb upstream region (as promoter accessibility is often
correlated with gene expression). We then count the number of fragments for each
cell that map to each of these regions, using the using the `FeatureMatrix()`
function. These steps are automatically performed by the `GeneActivity()`
function:

<!-- ```{r, message=FALSE, warning=FALSE} -->
<!-- gene.activities <- GeneActivity(sclc) -->
<!-- ``` -->

<!-- ```{r, message=FALSE, warning=FALSE} -->
<!-- # add the gene activity matrix to the Seurat object as a new assay and normalize it -->
<!-- sclc[['RNA']] <- CreateAssayObject(counts = gene.activities) -->
<!-- sclc <- NormalizeData( -->
<!--   object = sclc, -->
<!--   assay = 'RNA', -->
<!--   normalization.method = 'LogNormalize', -->
<!--   scale.factor = median(sclc$nCount_RNA) -->
<!-- ) -->
<!-- ``` -->


```{r, cache=FALSE}
# sclc <- subset(sclc, idents = 14, invert = TRUE)
sclc <- RenameIdents(
  object = sclc,
  '0' = 'Club cells-1',
  '1' = "NE-1",
  '2' = 'NE-2',
  '3' = 'NE-3',
  '4' = 'NE-4',
  '5' = 'NE-5',
  '6' = 'NE-6',
  '7' = 'Club cells-2',
  '8' = 'NFIB High NE',
  '9' = 'Club cells-3',
  '10' = 'NE-7',
  '11' = 'AT2 cells',
  '12' = 'NE-8',
  '13' = 'Ciliated cells',
  '14' = 'Endothelial cells',
  '15' = 'Stromal cells'
)
DimPlot(sclc, label = TRUE, repel = TRUE, reduction = "UMAP_joint") + NoLegend()

```



```{r message=FALSE, warning=FALSE, echo=FALSE}
saveRDS(object = sclc, file = "../data/sclc.rds")
```

</details>

<details>
  <summary>**Session Info**</summary>

```{r}
sessionInfo()
```

</details>