# Title     : Archetypes of Bulk SCLC Cell Line RNA-seq
# Objective : Determine the location in gene expression space of the archetypes and characterize which genes and gene sets are enriched in each archetype.
# Created by: smgroves
# Created on: 7/13/20


```{r install-packages echo = FALSE results = FALSE message = FALSE}
# install.packages('reticulate')
reticulate::use_condaenv("reticulate_PCHA", conda = "auto",
                         required = TRUE) # set TRUE to force R to use reticulate_PCHA
reticulate::py_discover_config("py_pcha")
# install.packages("BiocManager")
# BiocManager::install("vitkl/ParetoTI", dependencies = c("Depends", "Imports", "LinkingTo"))
library(cowplot)
library(ParetoTI)
library(ggplot2)
library(RColorBrewer)
library(reshape2)
library(factoextra)
library(ggfortify)
library(cluster)
set.seed(4355)

```
```{r preprocess-data}
# data <- '../data/CCLE_Minna_Thomas_COMBAT_vargenes.csv'
# data <- '/Users/smgroves/Dropbox (VU Basic Sciences)/SCLC_data/RNAseq/gdsc_minna_ccle/SCLC_CCLE_Minna_RNA-seq_batch-corrected_geneexp.csv'
data <- '../data/SCLC_combined_Minna_CCLE_batch_corrected_wo_lowgenes.csv'
x <- read.csv(data, header = TRUE, sep = ',', row.names = 1)
# drops <- c("t.S02353","t.S02297") #these are the two chromothripsis cases that are very different from the rest of the data points
# x <- x[,!(names(x) %in% drops)]
# 
# cm <- x[,!grepl("^t",colnames(x))]
# t <- x[,!grepl("^[cm]",colnames(x))]
# t <- expm1(t)
# tt <- transpose(t)
# rownames(tt) <- colnames(t)
# colnames(tt) <- rownames(t)
# 
# x <- cm
# x<- expm1(x)
xt <- transpose(x)
rownames(xt) <- colnames(x)
colnames(xt) <- rownames(x)
# data_labels <-'/Users/smgroves/Documents/MATLAB/particode/Data/combined_clusters_2020-05-27-MC.csv'
data_labels <-'/Users/smgroves/Documents/MATLAB/particode/Data/combined_clusters_2020-05-27-MC copy.csv'

# data_labels <-'../data/combined_clusters.csv'
labels <- read.csv(data_labels, header = TRUE, sep = ',', row.names = 1)
labels <- labels[row.names(xt),]

# labels_cm <- labels[row.names(xt),]

# labels_t <- labels[row.names(tt),]

## Dunn Index
# install.packages("clValid")
df2 <- xt[sample(nrow(xt),replace = TRUE),sample(ncol(xt), replace = TRUE)]

library(clValid)
intern <- clValid(df2, 2:6, clMethods=c("hierarchical"),
                  validation="internal")
summary(intern)
optimalScores(intern)
plot(intern)

#for background
df3 = scrambleDataFrame(df2, scrambling.rules = rules)
intern <- clValid(df3, 2:6, clMethods=c("hierarchical"),validation = 'internal')
summary(intern)

install.packages("remotes")
remotes::install_github("elbamos/clusteringdatasets")
library(clusteringdatasets)
b = make_blobs(n_samples = 120, n_features = 15950, centers = 1,
  cluster_std = 5, center_box = c(min(df2), max(df2)), shuffle = TRUE)
intern <- clValid(b$samples, 2:6, clMethods=c("hierarchical"),validation = 'internal')
summary(intern)
```

# PCA

Step 1 is to reduce the dimemsions of the data. We do this with a pca, which reduces dimensions linearly. We then visualize the number of PCs that describe the data well. 
```{r pca}
library(ggplot2)
# install.packages("viridis")
library(viridis)
pca <- prcomp(xt, center=TRUE, scale = FALSE)
# eig.val <- get_eigenvalue(pca)
# res.var <- get_pca_var(pca)
# res.var$contrib        # Contributions to the PCs

res.ind <- get_pca_ind(pca)
par(bg = "white")
# autoplot(pca) + theme_bw()

svg('pca_MYC.svg', width = 8, height = 6)
autoplot(pca, data = xt, colour = 'MYC', ) +theme_bw() +scale_color_viridis() 
dev.off()

pdf('pca_subtypes.pdf', width = 8, height = 6)
autoplot(pca, data =labels, colour = 'X10_08_2020', ) +theme_bw() 
dev.off()
options(repr.plot.width = 30, repr.plot.height = 6, repr.plot.res = 100)
fviz_eig(pca, ncp = 30, addlabels = TRUE)




svg('pca_variables.svg', width = 10, height = 10)

fviz_pca_var(pca, select.var = list(contrib = 20), xlim=rev(c(-3, 2)))
dev.off()

fviz_pca_ind(pca, col.ind="cos2")

# pred <- predict(pca, newdata=tt)
# expl.var <- round(pca$sdev^2/sum(pca$sdev^2)*100) # percent explained variance

###Plot result
# pc = c(1,2)
# COLOR <- c(1:6)
# PCH <- c(1,16)
# # png("pca_pred.png", units="in", width=5, height=4, res=200)
# op <- par(mar=c(4,4,1,1), ps=10)
# plot(pca$x[,pc], col=COLOR[labels_cm$phenotype], cex=PCH[1], 
#  xlab=paste0("PC ", pc[1], " (", expl.var[pc[1]], "%)"), 
#  ylab=paste0("PC ", pc[2], " (", expl.var[pc[2]], "%)")
# )
# points(pred[,pc], col=COLOR[labels_t$phenotype], cex=PCH[2])
# legend("topright", legend=levels(labels$phenotype), fill = COLOR, border=COLOR)
# legend("topleft", legend=c("training data", "validation data"), col=1, pch=PCH)
# par(op)
# dev.off()

x_pca <- t(pca$x)
x_pca <- x_pca[1:10,]

# t_pca <- t(pred)[1:10,]
```

```{r message= FALSE}
# install.packages('ICtest')
# library(ICtest)
# 
# PCAcov <- PCAasymp(xt, k=20)
# PCAcov
```

# Archetypes

```{r archetypes}
arc_rob = fit_pch_bootstrap(x_pca, n = 200, sample_prop = .8, seed = 2543, delta = 0,
                            noc = 5)
# arc_rob1 = fit_pch_bootstrap(x_pca, n = 200, sample_prop = .8, seed = 2543, delta = 1,
                            # noc = 5)
# arc_rob5 = fit_pch_bootstrap(x_pca, n = 200, sample_prop = .8, seed = 2543, delta = 5,
                            # noc = 5)
arc_ave <- average_pch_fits(arc_rob)
# arc_ave1 <- average_pch_fits(arc_rob1)
# arc_ave5 <- average_pch_fits(arc_rob5)
# 
# for (i in (0:8)){
#   arc_rob_mult = fit_pch_bootstrap(x_pca, n = 200, sample_prop = .8, seed = 2543, delta = i,
#                             noc = 5)
#   arc_ave_mult <- average_pch_fits(arc_rob_mult)
#   print(i)
#   print(arc_rob_mult)
#   print(arc_ave_mult)
# 
# }
arc<- fit_pch(x_pca, volume_ratio = 't_ratio',noc = 5)
# clusters = fit_pch(x_pca, noc = 5, method = "kmeans")

# plot_arc(arc_data = clusters, data = x_pca,
                   # which_dimensions = 1:2,
                   # data_lab = as.character(apply(clusters$S, 2, which.max))) +
            # ylim(-18, 17) + ggtitle("K-means clusters") + theme_bw()
# 
# p_pca = plot_arc(arc_data = arc_rob, data = t_pca, 
#                  which_dimensions = 1:3, line_size = 1.5,
#                  data_lab = as.numeric(t["ASCL1",]),
#                  text_size = 60, data_size = 6) 
# plotly::layout(p_pca, title = "ASCL1")

# p_pca = plot_arc(arc_data = arc_rob, data = x_pca, 
#                  which_dimensions = 1:3, line_size = 1.5,
#                  data_lab = as.character(labels$subtype),
#                  text_size = 60, data_size = 6) 
# plotly::layout(p_pca, title = "Subtype")



options(repr.plot.width = 7, repr.plot.height = 6, repr.plot.res = 100)
cols <- c(brewer.pal(6, "Set1"),'gray')
# cols <- c('red','blue','green',"purple",'grey','orange','black')
p_pca = plot_arc(arc_data = arc_ave, data = x_pca, 
                 which_dimensions = 1:3, line_size = 1.5, colors = cols,
                 data_lab = as.character(labels$subtype),
                 text_size = 60, data_size = 6) 
plotly::layout(p_pca, title = "Phenotype")

plot_arc(arc_data = arc_ave, data = x_pca,
                   which_dimensions = 1:2, colors = cols,
                  data_lab = as.character(labels$subtype)) + theme_bw()
# 
# options(repr.plot.width = 7, repr.plot.height = 6, repr.plot.res = 100)
# cols <- c(brewer.pal(6, "Set1"),'gray')
# 
# plot_arc(arc_data = arc_ave, data = x_pca,
#                    which_dimensions = 1:2, colors = cols,
#                   data_lab = as.character(labels$phenotype)) + theme_bw()
# 
# options(repr.plot.width = 20, repr.plot.height = 6, repr.plot.res = 100)
# plot_arc(arc_data = arc_ave, data = x_pca,
#                    which_dimensions = 1:2, colors = cols,
#                   data_lab = as.character(labels$phenotype)) + theme_bw()
# 
# plot_arc(arc_data = arc_ave, data = x_pca,colors = cols,
#                   data_lab = as.character(labels$phenotype),
#                    which_dimensions = 2:3) + theme_bw()
# 
# plot_arc(arc_data = arc_ave, data = x_pca,colors = cols,
#                   data_lab = as.character(labels$phenotype),
#                    which_dimensions = 3:4) + theme_bw()
#           # align = "vh", nrow = 1) 

```



```{r randomize-archetypes}
# use permutations within each dimension - this is only possible for less than 8 vertices because computing convex hull gets exponentially slower with more dimensions
start = Sys.time()
pch_rand = randomise_fit_pch(x_pca, arc_data = arc_ave,
                             n_rand = 1000,
                             replace = FALSE, 
                             volume_ratio = "t_ratio",
                             maxiter = 500, delta = 1, conv_crit = 1e-4)
# use type m to run on a single machine or cloud
# type = "m", clust_options = list(cores = 3))
# use clustermq (type cmq) to run as jobs on a computing cluster (higher parallelisation)
# type = "cmq", clust_options = list(njobs = 10)) 

# This analysis took:
Sys.time() - start
# plot(pch_rand, type = c("variance_ratio"), nudge_y = 5)


```

```{r reconstruct-from-pca}
reconstruct_from_pca <- function(data_train, data_output){
    mu = colMeans(data_train)
    pca <- prcomp(data_train, center=TRUE)
    nComp = 10
    if (is.null(data_output)){Xhat = pca$x[,1:nComp] %*% t(pca$rotation[,1:nComp])}else
    {Xhat = data_output[,1:nComp] %*% t(pca$rotation[,1:nComp])}
    Xhat = scale(Xhat, center = -mu, scale = FALSE)
    return(Xhat)}

arc_genespace <- reconstruct_from_pca(xt, data_output = t(arc_ave$XC))
write.csv(t(arc_genespace), 'arc_gene-space-linear.csv')
```


```{r}
library(matrixStats)

activ = measure_activity(as.matrix(x), activity_method = 'pseudoinverse',# row names are assumed to be gene identifiers,
                         which = 'BP', return_as_matrix = F,
                         taxonomy_id = 9606, keytype = "ALIAS",
                         lower = 10, upper = 1000)
save(activ, file="../int/activ_bulk.Robj")

activ <- within(activ, rm('2__deoxyribonucleotide_biosynthetic_process','2__deoxyribonucleotide_metabolic_process','2_oxoglutarate_metabolic_process','3__phosphoadenosine_5__phosphosulfate_metabolic_process',
'3__UTR_mediated_mRNA_destabilization',
'3__UTR_mediated_mRNA_stabilization',
'7_methylguanosine_mRNA_capping',
'7_methylguanosine_RNA_capping',
'4_hydroxyproline_metabolic_process'))

#%%


activ <- within(activ, rm(`_de_novo__posttranslational_protein_folding`,
`_de_novo__protein_folding`,
`poly_A_+_mRNA_export_from_nucleus`))

#%%

rownames(x)<-gsub("-", ".", rownames(x))


#%%

ix <- which(rownames(x) %in%c('1.Sep','10.Sep','11.Mar','11.Sep','2.Sep','3.Mar','3.Sep','4.Mar','4.Sep','5.Sep', '5.Mar','6.Sep','6.Mar','7.Sep','7.Mar','8.Sep','8.Mar','9.Sep','9.Mar'))
clean <- x[-ix, ]



data_attr = merge_arch_dist(arc_data = arc_ave, data = x_pca, 
                            feature_data = as.matrix(x),
                            colData = activ, 
                            dist_metric = c("euclidean", "arch_weights")[1],
                            colData_id = "cells", rank = F) 

enriched_genes = find_decreasing_wilcox(data_attr$data, data_attr$arc_col,
                                features = data_attr$features_col,
                                bin_prop = 0.05, method = "BioQC")
# write.csv(enriched_genes, '../data/enriched-genes.csv')

enriched_sets = find_decreasing_wilcox(data_attr$data, data_attr$arc_col,
                                features = data_attr$colData_col,
                                bin_prop = 0.1, method = "BioQC")
# write.csv(enriched_sets, '../data/enriched_sets.csv')

labs = get_top_decreasing(summary_genes = enriched_genes, summary_sets = enriched_sets,
                          cutoff_genes = 0.05,cutoff_sets = 0.05,
                          cutoff_metric = "wilcoxon_p_val", 
                          p.adjust.method = "none", 
                          order_by = "mean_diff", order_decreasing = T,
                          min_max_diff_cutoff_g = 0.05, min_max_diff_cutoff_f  = 0.05)
p_pca = plot_arc(arc_data = arc_ave, data = x_pca,
                 which_dimensions = 1:3, line_size = 1.5,
                 data_lab = activ_alias$regulation_of_protein_localization_to_chromosome__telomeric_region,
                 text_size = 60, data_size = 6)
plotly::layout(p_pca, title = "regulation_of_protein_localization_to_chromosome__telomeric_region activity")


enriched_genes_gam = find_decreasing(data_attr$data, data_attr$arc_col,
                                features = data_attr$features_col, return_only_summary = TRUE)
write.csv(enriched_genes_gam, '../data/enriched_genes_gam_COMBAT.csv')

enriched_sets_gam = find_decreasing(data_attr$data, data_attr$arc_col,
                                features = data_attr$colData_col, return_only_summary = TRUE)
write.csv(enriched_sets_gam, '../data/enriched_sets_gam_COMBAT.csv')

labs_gam = get_top_decreasing(summary_genes = enriched_genes_gam, summary_sets = enriched_sets_gam,
                          cutoff_genes = 0.05,cutoff_sets = 0.05,
                          cutoff_metric = "mean_prob", 
                          p.adjust.method = "none", 
                          order_by = "deriv50", order_decreasing = F,
                          min_max_diff_cutoff_g = 1)
save(labs_gam, file="../int/labs_gam.Robj")


```







